### ISSUE TITLE
Missing implementations for JVP and VMAP

### ISSUE THREAD

**Issue by ASKabalan** [DifferentiableUniverseInitiative/jaxDecomp#52]
A few missing rules are needed for gradient computation in ODE integration 

Like the JVP of the ODETerm to get the Adjoint Term

```python
import jax.numpy as jnp
import jax.numpy as jnp; import jax
from jaxdecomp.fft import pfft3d
from jax import custom_jvp
from functools import partial
from jaxdecomp._src.jax.fft import JAXFFTPrimitive , ADJOINT
from jax.lax import FftType

def f(a):
  return pfft3d(a).real.sum()

a = jnp.arange(4*4*4).reshape(4,4,4).astype(jnp.float32)

jax.grad(f)(a)  

#jax.jacrev(f)(a) # Won't work needs batching rule

#jax.jacfwd(f)(a) # Won't work needs jvp rule

@partial(custom_jvp, nondiff_argnums=(1, 2))
def pfft_jvp(x, fft_type, adjoint = False) :

  return JAXFFTPrimitive.outer_lowering(x, fft_type=fft_type, adjoint=adjoint)

@pfft_jvp.defjvp
def pfft_jvp_rule(fft_type, adjoint , primals, tangents):

  x, = primals
  t, = tangents
  primal_out = pfft_jvp(x, fft_type=fft_type, adjoint=adjoint)
  tanget_out = pfft_jvp(t, fft_type=ADJOINT(fft_type), adjoint=not adjoint)
  return primal_out, tanget_out

def pfft3d_jvp(x, adjoint = False):
  return pfft_jvp(x, fft_type = FftType.FFT, adjoint = adjoint)

def f2(a):
  return pfft3d_jvp(a).real.sum()

# jax.jacfwd(f2)(a) # Wont work need transposing rule
```

Custom transpose and batching rule must be implemented

It is unclear to me in what order, since you need a vjp rule for a custom vmap rule as explained here

https://github.com/jax-ml/jax/pull/25259

Also there is a possible API to have both forward and backward rules 

https://github.com/jax-ml/jax/pull/22457

For now I will implement only jvp and custom transpose rule

waiting for JAX team answer for my discussion 

https://github.com/jax-ml/jax/discussions/25732
